test/e2e/upgrade_test_test.go

```diff
@@ -0,0 +1,95 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+	"golang.org/x/net/context"
+)
+
+const (
+	upgradeNamespace = "upgrade-workload"
+)
+
+// runKibishiiTests runs kibishii tests on the provider.
+func runUpgradeTests(client testClient, toVeleroImage, providerName, veleroCLI, veleroNamespace, backupName, restoreName, backupLocation string,
+	useVolumeSnapshots bool, registryCredentialFile string) error {
+	oneHourTimeout, _ := context.WithTimeout(context.Background(), time.Minute*60)
+	serviceAccountName := "default"
+	if err := kibishiiPrepareBeforeBackup(client, providerName, oneHourTimeout, upgradeNamespace, serviceAccountName, registryCredentialFile); err != nil {
+		return errors.Wrapf(err, "Failed to install and prepare data for kibishii %s", upgradeNamespace)
+	}
+
+	defer func() {
+		if err := deleteNamespace(oneHourTimeout, client, upgradeNamespace, true); err != nil {
+			fmt.Println(errors.Wrapf(err, "failed to delete the namespace %q", upgradeNamespace))
+		}
+	}()
+
+	if err := veleroBackupNamespace(oneHourTimeout, veleroCLI, veleroNamespace, backupName, upgradeNamespace, backupLocation, useVolumeSnapshots); err != nil {
+		veleroBackupLogs(oneHourTimeout, veleroCLI, veleroNamespace, backupName)
+		return errors.Wrapf(err, "Failed to backup kibishii namespace %s", upgradeNamespace)
+	}
+
+	if providerName == "vsphere" && useVolumeSnapshots {
+		// Wait for uploads started by the Velero Plug-in for vSphere to complete
+		// TODO - remove after upload progress monitoring is implemented
+		fmt.Println("Waiting for vSphere uploads to complete")
+		if err := waitForVSphereUploadCompletion(oneHourTimeout, time.Hour, upgradeNamespace); err != nil {
+			return errors.Wrapf(err, "Error waiting for uploads to complete")
+		}
+	}
+	fmt.Printf("Simulating a disaster by removing namespace %s\n", upgradeNamespace)
+	if err := deleteNamespace(oneHourTimeout, client, upgradeNamespace, true); err != nil {
+		return errors.Wrapf(err, "failed to delete namespace %s", upgradeNamespace)
+	}
+	var cli string
+	var err error
+	main_tag := "main"
+	upgradeToVersion := veleroImage[strings.Index(toVeleroImage, ":")+1:]
+	if upgradeToVersion == main_tag {
+		cli = veleroCLI
+	} else {
+		cli, err = installVeleroCLI(upgradeToVersion)
+		if err != nil {
+			return errors.Wrapf(err, "Fail to install velero CLI.")
+		}
+	}
+	if err := veleroInstall(context.Background(), cli, toVeleroImage, veleroNamespace, cloudProvider, objectStoreProvider, useVolumeSnapshots,
```

Comment:
@zubron Yes, installation function used in e2e applys the CRDs first just like the instructions in the guide https://github.com/vmware-tanzu/velero/blob/main/test/e2e/install.go#L153

And I'm curious if we just replace the velero server image as the guide said, are all necessary resources installed? For example, if we introduce a new configmap which is necessary to the new velero server

Is it OK to just apply all new resources generated by `velero install --dry-run`?

Reply:
I didn't realize that the new install code already applied the CRDs like that, so you can ignore my comment :)

I don't think we need to replace it, as long as the new install code replaces the CRDs and the images, it's fine. Long term, I think we want to improve the user experience so that they can just type something like `velero upgrade` and that will take care of performing all the necessary steps.

For your question about introducing something like a new `ConfigMap`, I think in that case we would need to update our instructions to match that. The current instructions definitely show that we should improve this process 👍

Reply:
Thanks all for discussion and clarification!